File: sdk/typescript/src/turn-status.ts
import { TurnStatus } from "./generated/models/TurnStatus.js";

export { TurnStatus };

export const TERMINAL_TURN_STATUSES: readonly TurnStatus[] = [
  TurnStatus.Completed,
  TurnStatus.Incomplete,
  TurnStatus.Failed,
  TurnStatus.Cancelled,
] as const;

const TERMINAL = new Set<TurnStatus>(TERMINAL_TURN_STATUSES);

export const ACTIVE_TURN_STATUSES: readonly TurnStatus[] = Object.values(TurnStatus)
  .filter((status) => !TERMINAL.has(status));

export function isTerminalTurnStatus(status: TurnStatus): boolean {
  return TERMINAL.has(status);
}

export function isTurnOver(value: TurnStatus | { status: TurnStatus; terminal?: boolean }): boolean {
  return typeof value === "string"
    ? isTerminalTurnStatus(value)
    : value.terminal ?? isTerminalTurnStatus(value.status);
}

File: sdk/rust/src/turn_status.rs
//! Turn lifecycle classification shared by facade and stream reducers.

use crate::models;

pub const ALL_TURN_STATUSES: [models::TurnStatus; 8] = [
    models::TurnStatus::Queued,
    models::TurnStatus::Running,
    models::TurnStatus::Waiting,
    models::TurnStatus::BudgetHold,
    models::TurnStatus::Completed,
    models::TurnStatus::Incomplete,
    models::TurnStatus::Failed,
    models::TurnStatus::Cancelled,
];

pub const TERMINAL_TURN_STATUSES: [models::TurnStatus; 4] = [
    models::TurnStatus::Completed,
    models::TurnStatus::Incomplete,
    models::TurnStatus::Failed,
    models::TurnStatus::Cancelled,
];

pub fn is_terminal_status(status: models::TurnStatus) -> bool {
    TERMINAL_TURN_STATUSES.contains(&status)
}

pub fn is_turn_over(change: &models::TurnChange) -> bool {
    change.terminal || is_terminal_status(change.status)
}
